db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ProjectRestrictNode.java

```diff
@@ -809,6 +809,17 @@ else if(!(childResult instanceof FromBaseTable)){
         }
         accessPathModified=true;
 
+        if (childResult.getResultColumns().isFromExprIndex()) {
+            /* We get a shallow copy of the ResultColumnList and its
+             * ResultColumns.  (Copy maintains ResultColumn.expression for now.)
+             */
+            resultColumns = childResult.getResultColumns().copyListAndObjects();
```

Comment:
I have a concern with the replacement of the `resultColumns` here. As all the column references that are on top of this ProjectRestrictNode that reference the result columns in `resultColumns` will be broken. I'm trying to use s derived table as a case to demonstrate the problem, but currently in the presence of derived table, index are always qualified as a non-covering index, so this logic is not triggered. We need to check back here once the derived table issue is fixed.

Reply:
After two days debugging with cases including derived tables and outer joins, I believe that breaking column mapping is unfortunately inevitable. Main reason is that if `childResult` comes from an index on expressions, the number of RCs may change. Example could be here we have two result columns for A and B, but the index expression is A + B and the `childResult` has only one result column. Or we could have only column A here, but actually we have multiple index expressions defined on A and referenced like A + 1 and A * 3.

Once we enter this code path, parent nodes must have their `resultColumns` rebuilt, too. Good news is that this propagation stops at top-level select of current query block. There we don't have to make new RC instances, but just replace their expressions.

The outer join failure you discovered is a combined one. Fixing it requires 1) expression collection is fixed, 2) result column rebuilding is fixed, and 3) `indexExpression` of RCs must be set and propagated properly. The third item is a newly found issue.

Reply:
Based on our discussion yesterday, I made two changes to the code:

1. not propagating `fromIndexExpr` flag across query block boundary
2. when rebuilding result columns of join node, it rebuilds only those from index expressions

With these two changes, all tests work with no need of `UpdateColumnReferenceVisitor`. Thus, I remove it and revert unnecessary changes in the latest commit.